File: src/main/java/stdweb/rest/AccountController.java
package stdweb.rest;

import org.ethereum.core.Account;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.RestController;
import stdweb.Core.AddressDecodeException;
import stdweb.Core.Convert2json;
import stdweb.Core.Utils;
import stdweb.Ledger.AccountStore;
import stdweb.Ledger.LedgerAccount;
import stdweb.Ledger.LedgerQuery;
import stdweb.Ledger.LedgerStore;

import javax.servlet.http.HttpServletRequest;
import java.io.IOException;
import java.math.BigDecimal;
import java.sql.SQLException;

import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
import static org.springframework.web.bind.annotation.RequestMethod.GET;
import static stdweb.Core.Utils.address_decode;

/**
 * Created by bitledger on 13.11.15.
 */

@RestController
public class AccountController {

    AccountStore astore;
    @RequestMapping(value = "/account/{accountId}/{offset}", method = GET, produces = APPLICATION_JSON_VALUE)
    @ResponseBody

    public String getAccountLedger(@PathVariable String accountId,@PathVariable String offset,HttpServletRequest request) throws IOException {

        try {
            long t1=System.currentTimeMillis();
            LedgerStore ledgerStore = LedgerStore.getLedgerStore();
            LedgerQuery ledgerQuery = LedgerQuery.getQuery(ledgerStore);

            JSONArray jsonArray = ledgerQuery.LedgerSelect(accountId, offset);

            Utils.log("LedgerSql", t1, request, true);
            //JSONObject entriesJson=new JSONObject();

            long t2=System.currentTimeMillis();
            JSONObject entriesJson=ledgerQuery.acc_entry_count(accountId,offset);

            Utils.log("entries count",t2,request,true);

            byte[] acc=address_decode(accountId);
            LedgerAccount account=astore.get(acc);

            t2=System.currentTimeMillis();
            //BigDecimal ledgerBlockBalance = ledgerQuery.getLedgerAccountBalance(account,ledgerQuery.getSqlTopBlock());
            BigDecimal ledgerBlockBalance = account!=null ? account.getBalance() : BigDecimal.ZERO;

            //Utils.log("acc balance",t2,request,true);


            //BigDecimal ledgerBlockBalance=BigDecimal.ZERO;
            entriesJson.put("balance", Convert2json.BD2ValStr(ledgerBlockBalance, true));
            if (account!=null)
                entriesJson.put("addresstype",account.isContract() ? "Contract" : "Account");


            entriesJson.put("entries",jsonArray);
            String s=entriesJson.toJSONString();



            s=s.replace(":"," ");
            Utils.log("AccountLedger",t1,request);
            return s;

        } catch (SQLException e) {
            e.printStackTrace();
            return  e.toString();
        } catch (AddressDecodeException e) {
            e.printStackTrace();
            return  e.toString();
        }
    }

    public AccountController() throws SQLException {
        astore=AccountStore.getInstance();
    }
}

File: src/main/java/stdweb/rest/AccountController.kt
package stdweb.rest


import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import java.io.IOException
import java.math.BigDecimal
import java.sql.SQLException
import javax.servlet.http.HttpServletRequest


import org.springframework.http.MediaType.APPLICATION_JSON_VALUE
import org.springframework.web.bind.annotation.*
import org.springframework.web.bind.annotation .RequestMethod.GET
import stdweb.Core.*
import stdweb.Core.Utils.address_decode
import stdweb.Entity.LedgerAccount
import stdweb.Entity.LedgerEntry
import stdweb.Repository.LedgerAccountRepository
import stdweb.Repository.LedgerBlockRepository
import stdweb.Repository.LedgerEntryRepository
import java.util.*

/**
 * Created by bitledger on 13.11.15.
 */

@RestController
class AccountController{

    @Autowired
    var ledgRepo : LedgerEntryRepository? = null

    @Autowired
    var blockRepo : LedgerBlockRepository? = null


    @Autowired
    var acckRepo : LedgerAccountRepository? = null

    @RequestMapping(value = "/account/{accountId}/{page}", method = arrayOf( RequestMethod.GET), produces = arrayOf(MediaType.APPLICATION_JSON_VALUE))
    @ResponseBody
    fun getAccountLedger(@PathVariable accountId: String, @PathVariable page: String, request: HttpServletRequest)
            : HashMap<String,Any> {

        val res: ResponseEntity<LedgerEntry>
        val t1 = System.currentTimeMillis()
        res = ResponseEntity(null, HttpStatus.NOT_IMPLEMENTED)

        val addr=address_decode(accountId)
        val acc   = acckRepo?.findByAddress(addr)
        val offs=(page.toInt()-1)*25
        val zeroAccount= LedgerAccount(Utils.ZERO_BYTE_ARRAY_20)

        //val content = ledgRepo?.getAccountLedgerPage(acc!!.id,offs) ?:  ArrayList<LedgerEntry>()
        //val content = ledgRepo?.getAccountLedger(acc!!.id,offs,page.toInt()*25) ?:  ArrayList<LedgerEntry>()
        val content = ledgRepo?.getAccountLedger(acc!!.id,offs,page.toInt()*25) ?:  ArrayList<LedgerEntry>()

        val result  = ArrayList<LedgerEntry>()

        content.forEach {
            result.add(it)
            if (it.fee.signum()==1)
                with (LedgerEntry()) {
                    account             = it.account
                    tx                  = it.tx
                    offsetAccount       = zeroAccount
                    amount              = it.fee.negate()
                    block               = it.block
                    blockTimestamp      = it.blockTimestamp
                    depth               = it.depth
                    gasUsed             = 0
                    entryType           = EntryType.TxFee
                    fee                 = BigDecimal.ZERO
                    grossAmount         = this.fee.negate()
                    entryResult         = EntryResult.Ok

                    balance             =it.balance
                    it.balance          =it.balance.add(it.fee)
                    result.add(this)
                }
        }

        val ret=HashMap<String,Any>()


        val firstBlock  =blockRepo?.findOne(ledgRepo?.getFirstAccountBlock(acc?.id ))
        val lastBlock   =blockRepo?.findOne(ledgRepo?.getLastAccountBlock(acc?.id ))

        val entriesCount=ledgRepo?.getEntriesCount(acc!!.id) ?: 0
        with(ret){
            put("entries_count", entriesCount)
            put("page_count",entriesCount/25+1)
            put("balance",Convert2json.BD2ValStr(acc?.balance,false))
            put("addresstype",if (acc?.isContract ?: false) "Contract" else "account")
            put("entries", result)
            put("firstblock","${firstBlock?.id}, Date ${firstBlock?.BlockDateTime}")
            put("lastblock","${lastBlock?.id}, Date ${lastBlock?.BlockDateTime}")
        }
        Utils.log("AccountLedger", t1, request, res)

        return ret
    }


}


